src/ralph/utils.py

```diff
@@ -171,39 +186,80 @@ def statements_are_equivalent(statement_1: dict, statement_2: dict) -> bool:
-            yield json.loads(raw_document)
-        except (TypeError, json.JSONDecodeError) as error:
-            msg = "Failed to decode JSON: %s, for document: %s"
+            yield parser(raw_document)
+        except exceptions as error:
+            msg = "Failed to decode JSON: %s, for document: %s, at line %s"
             if ignore_errors:
-                logger_class.warning(msg, error, raw_document)
+                logger_class.warning(msg, error, raw_document, i)
                 continue
-            logger_class.error(msg, error, raw_document)
-            raise BackendException(msg % (error, raw_document)) from error
+            logger_class.error(msg, error, raw_document, i)
+            raise BackendException(msg % (error, raw_document, i)) from error
 
 
-def read_raw(
+def parse_dict_to_bytes(
     documents: Iterable[Dict[str, Any]],
     encoding: str,
     ignore_errors: bool,
     logger_class: logging.Logger,
 ) -> Iterator[bytes]:
     """Read the `documents` Iterable with the `encoding` and yield bytes."""
-    for document in documents:
+    for i, document in enumerate(documents):
+        try:
+            yield f"{json.dumps(document)}\n".encode(encoding)
+        except (TypeError, ValueError) as error:
+            msg = "Failed to encode JSON: %s, for document: %s, at line %s"
+            if ignore_errors:
+                logger_class.warning(msg, error, document, i)
+                continue
+            logger_class.error(msg, error, document, i)
+            raise BackendException(msg % (error, document, i)) from error
+
+
+async def async_parse_dict_to_bytes(
+    documents: AsyncIterable[Dict[str, Any]],
+    encoding: str,
+    ignore_errors: bool,
+    logger_class: logging.Logger,
+) -> AsyncIterator[bytes]:
+    """Read the `documents` Iterable with the `encoding` and yield bytes."""
+    i = 0
+    async for document in documents:
         try:
-            yield json.dumps(document).encode(encoding)
+            yield f"{json.dumps(document)}\n".encode(encoding)
         except (TypeError, ValueError) as error:
-            msg = "Failed to convert document to bytes: %s"
+            msg = "Failed to encode JSON: %s, for document: %s, at line %s"
             if ignore_errors:
-                logger_class.warning(msg, error)
+                logger_class.warning(msg, error, document, i)
                 continue
-            logger_class.error(msg, error)
-            raise BackendException(msg % error) from error
+            logger_class.error(msg, error, document, i)
+            raise BackendException(msg % (error, document, i)) from error
+
+        i += 1
+
+
+def iter_by_batch(data: Iterable[Any], batch_size: int):
+    """Iterate over `data` Iterable and yield batches of size `batch_size`."""
+    batch = []
+    for document in data:
+        batch.append(document)
+        if len(batch) >= batch_size:
+            yield batch
+            batch = []
+    if batch:
+        yield batch
```

Comment:
Do we have to yield a list? Else we could use the `itertools` implementation that will be added in Python 3.12 (https://docs.python.org/3/library/itertools.html#itertools.batched)
```suggestion
def batched(iterable: Iterable, n: int):
    """
    Batch data from an iterable into tuples of length n. The last batch may be
    shorter than n.

    Taken from itertools.batched introduced in python 3.12
    """
    if n < 1:
        raise ValueError("n must be at least one")
    it = iter(iterable)
    while batch := tuple(islice(it, n)):
        yield batch
```
WDYT?

Reply:
Great idea) Thanks) Have updated the implementation as you suggested:

```python
def iter_by_batch(iterable: Iterable, n: int) -> Iterable[list]:
    """Batch `iterable` into lists of length `n`. The last batch may be less than `n`.

    Taken from itertools.batched introduced in python 3.12
    """
    if n < 1:
        raise ValueError("n must be at least one")

    it = iter(iterable)
    while batch := list(islice(it, n)):
        yield batch
```
There are two slight differences :
- Could we keep yielding lists (it seems the Mongo data backend relies on the output being a list)?
- Could we postpone the method name change in a dedicated PR (to avoid rebase conflicts with other PRs)?

Reply:
Fine with this implementation! And we can stick with this name too!
I added a small remark on the docstring though

Reply:
Thanks) Updated it)